File: backend/app/services/fitting/cpmg_reduction.py
from __future__ import annotations

import math
from typing import Any, Dict, List, Optional, Tuple
import numpy as np


def compute_r2eff_profile(
    ncycs: List[float],
    intensities: List[float],
    uncertainties: Optional[List[float]],
    time_t2: float,
) -> Dict[str, Any]:
    """
    Compute nu_cpmg, R2eff and propagated errors from raw intensity profile.
    Plane with ncyc == 0 is the reference plane (I0).
    """
    if time_t2 <= 0:
        raise ValueError("time_t2 must be positive")

    # Find reference plane
    ref_idx = None
    for idx, ncyc in enumerate(ncycs):
        if abs(ncyc) < 1e-6:
            ref_idx = idx
            break

    if ref_idx is None:
        return {
            "error": "Missing reference plane (ncyc = 0)",
            "nu_cpmg": [],
            "r2eff": [],
            "r2eff_err": [],
            "valid": False,
        }

    i0 = intensities[ref_idx]
    if i0 <= 0:
        return {
            "error": f"Non-positive reference intensity: {i0}",
            "nu_cpmg": [],
            "r2eff": [],
            "r2eff_err": [],
            "valid": False,
        }

    i0_err = uncertainties[ref_idx] if uncertainties and len(uncertainties) > ref_idx else (0.02 * i0)

    raw_points = []
    invalid_points: List[int] = []
    for idx, (ncyc, i_val) in enumerate(zip(ncycs, intensities)):
        if abs(ncyc) < 1e-6:
            continue  # Skip reference plane in R2eff dispersion curve

        nu = ncyc / time_t2
        if i_val <= 0:
            invalid_points.append(idx)
            continue

        ratio = i_val / i0
        r2eff = -math.log(ratio) / time_t2

        i_err = uncertainties[idx] if uncertainties and len(uncertainties) > idx else (0.02 * i_val)
        # Error propagation: sigma_R2 = (1 / time_t2) * sqrt((sigma_I / I)^2 + (sigma_I0 / I0)^2)
        rel_err_sq = (i_err / i_val) ** 2 + (i0_err / i0) ** 2
        r2eff_err = math.sqrt(rel_err_sq) / time_t2

        raw_points.append((nu, r2eff, r2eff_err))

    # Always sort points ascending by CPMG frequency
    raw_points.sort(key=lambda p: p[0])
    nu_list = [p[0] for p in raw_points]
    r2eff_list = [p[1] for p in raw_points]
    err_list = [p[2] for p in raw_points]

    return {
        "nu_cpmg": nu_list,
        "r2eff": r2eff_list,
        "r2eff_err": err_list,
        "i0": i0,
        "i0_err": i0_err,
        "invalid_points": invalid_points,
        "valid": len(nu_list) > 0,
    }


def compute_rex_and_flatness(
    nu_cpmg: List[float],
    r2eff: List[float],
    r2eff_err: List[float],
) -> Dict[str, Any]:
    """
    Compute Rex = R2eff(nu_min) - R2eff(nu_max) and chi2 flatness score against horizontal line.
    Averages duplicate points at the minimum and maximum frequencies.
    """
    if len(nu_cpmg) < 2 or len(r2eff) < 2:
        return {
            "rex": 0.0,
            "rex_err": 0.0,
            "chi2_red": 0.0,
            "is_flat": True,
        }

    # Sort by nu_cpmg
    points = sorted(zip(nu_cpmg, r2eff, r2eff_err), key=lambda p: p[0])
    nu_sorted = [p[0] for p in points]
    r2_sorted = [p[1] for p in points]
    err_sorted = [p[2] if p[2] > 1e-6 else 1.0 for p in points]

    min_nu = nu_sorted[0]
    max_nu = nu_sorted[-1]

    # Collect points at min_nu and max_nu (accounting for possible duplicates)
    min_pts = [(r, e) for n, r, e in points if abs(n - min_nu) < 1e-3]
    max_pts = [(r, e) for n, r, e in points if abs(n - max_nu) < 1e-3]

    r2_min = float(np.mean([p[0] for p in min_pts]))
    err_min = math.sqrt(sum(p[1] ** 2 for p in min_pts)) / len(min_pts)

    r2_max = float(np.mean([p[0] for p in max_pts]))
    err_max = math.sqrt(sum(p[1] ** 2 for p in max_pts)) / len(max_pts)

    rex = r2_min - r2_max
    rex_err = math.sqrt(err_min ** 2 + err_max ** 2)

    # Flatness test: weighted mean and chi-square
    weights = [1.0 / (e ** 2) for e in err_sorted]
    sum_w = sum(weights)
    weighted_mean_r2 = sum(w * r for w, r in zip(weights, r2_sorted)) / sum_w if sum_w > 0 else float(np.mean(r2_sorted))

    chi2 = sum(w * ((r - weighted_mean_r2) ** 2) for w, r in zip(weights, r2_sorted))
    dof = max(1, len(r2_sorted) - 1)
    chi2_red = chi2 / dof

    return {
        "rex": max(0.0, rex),
        "rex_raw": rex,
        "rex_err": rex_err,
        "chi2_red": chi2_red,
        "is_flat": chi2_red < 1.5,
        "weighted_mean_r2": weighted_mean_r2,
    }


def estimate_delta_omega_fast_exchange(
    rex: float,
    kex: float,
    pb: float,
    b0_mhz: float,
    xi_ratio: float = 0.101329118,
) -> float:
    """
    Estimate unsigned |dw| (ppm) from Rex in the fast-exchange limit:
    Rex = pa * pb * dw^2 / kex
    => |dw_rad_s| = sqrt(Rex * kex / (pa * pb))
    => |dw_ppm| = |dw_rad_s| / (2 * pi * B0 * xi)
    """
    if rex <= 0 or kex <= 0 or pb <= 0 or pb >= 1.0 or b0_mhz <= 0:
        return 0.0

    pa = 1.0 - pb
    denom = pa * pb
    if denom <= 1e-9:
        return 0.0

    dw_rad_s = math.sqrt((rex * kex) / denom)
    nu0_mhz = b0_mhz * xi_ratio
    dw_ppm = dw_rad_s / (2.0 * math.pi * nu0_mhz)
    return round(dw_ppm, 3)

File: frontend/src/lib/cpmgReduction.ts
export interface R2effResult {
  nu_cpmg: number[];
  r2eff: number[];
  r2eff_err: number[];
  i0: number;
  i0_err: number;
  invalid_points: number[];
  valid: boolean;
  error?: string;
}

export interface RexFlatnessResult {
  rex: number;
  rex_raw: number;
  rex_err: number;
  chi2_red: number;
  is_flat: boolean;
  weighted_mean_r2: number;
}

export function computeR2effProfile(
  ncycs: number[],
  intensities: number[],
  uncertainties: number[] | undefined,
  time_t2: number
): R2effResult {
  if (time_t2 <= 0) {
    return {
      error: "time_t2 must be positive",
      nu_cpmg: [],
      r2eff: [],
      r2eff_err: [],
      i0: 0,
      i0_err: 0,
      invalid_points: [],
      valid: false,
    };
  }

  // Find reference plane (ncyc == 0)
  let ref_idx: number | null = null;
  for (let i = 0; i < ncycs.length; i++) {
    if (Math.abs(ncycs[i]) < 1e-6) {
      ref_idx = i;
      break;
    }
  }

  if (ref_idx === null) {
    return {
      error: "Missing reference plane (ncyc = 0)",
      nu_cpmg: [],
      r2eff: [],
      r2eff_err: [],
      i0: 0,
      i0_err: 0,
      invalid_points: [],
      valid: false,
    };
  }

  const i0 = intensities[ref_idx];
  if (i0 <= 0) {
    return {
      error: `Non-positive reference intensity: ${i0}`,
      nu_cpmg: [],
      r2eff: [],
      r2eff_err: [],
      i0: 0,
      i0_err: 0,
      invalid_points: [],
      valid: false,
    };
  }

  const i0_err = uncertainties && uncertainties.length > ref_idx ? uncertainties[ref_idx] : 0.02 * i0;

  const nu_list: number[] = [];
  const r2eff_list: number[] = [];
  const err_list: number[] = [];
  const invalid_points: number[] = [];

  for (let idx = 0; idx < ncycs.length; idx++) {
    const ncyc = ncycs[idx];
    if (Math.abs(ncyc) < 1e-6) continue; // Skip reference plane

    const i_val = intensities[idx];
    if (i_val <= 0) {
      invalid_points.push(idx);
      continue;
    }

    const nu = ncyc / time_t2;
    const ratio = i_val / i0;
    const r2eff = -Math.log(ratio) / time_t2;

    const i_err = uncertainties && uncertainties.length > idx ? uncertainties[idx] : 0.02 * i_val;
    const rel_err_sq = Math.pow(i_err / i_val, 2) + Math.pow(i0_err / i0, 2);
    const r2eff_err = Math.sqrt(rel_err_sq) / time_t2;

    nu_list.push(nu);
    r2eff_list.push(r2eff);
    err_list.push(r2eff_err);
  }

  return {
    nu_cpmg: nu_list,
    r2eff: r2eff_list,
    r2eff_err: err_list,
    i0,
    i0_err,
    invalid_points,
    valid: nu_list.length > 0,
  };
}

export function computeRexAndFlatness(
  nu_cpmg: number[],
  r2eff: number[],
  r2eff_err: number[]
): RexFlatnessResult {
  if (nu_cpmg.length < 2 || r2eff.length < 2) {
    return {
      rex: 0.0,
      rex_raw: 0.0,
      rex_err: 0.0,
      chi2_red: 0.0,
      is_flat: true,
      weighted_mean_r2: r2eff[0] || 0.0,
    };
  }

  // Combine and sort by nu_cpmg
  const points = nu_cpmg.map((nu, i) => ({
    nu,
    r2: r2eff[i],
    err: r2eff_err[i] && r2eff_err[i] > 1e-6 ? r2eff_err[i] : 1.0,
  })).sort((a, b) => a.nu - b.nu);

  const r2_min = points[0].r2;
  const err_min = points[0].err;
  const r2_max = points[points.length - 1].r2;
  const err_max = points[points.length - 1].err;

  const rex_raw = r2_min - r2_max;
  const rex_err = Math.sqrt(err_min * err_min + err_max * err_max);

  // Weighted mean for flatness test
  let sum_w = 0;
  let sum_wr = 0;
  for (const p of points) {
    const w = 1.0 / (p.err * p.err);
    sum_w += w;
    sum_wr += w * p.r2;
  }
  const weighted_mean_r2 = sum_w > 0 ? sum_wr / sum_w : r2_min;

  let chi2 = 0;
  for (const p of points) {
    const w = 1.0 / (p.err * p.err);
    chi2 += w * Math.pow(p.r2 - weighted_mean_r2, 2);
  }
  const dof = Math.max(1, points.length - 1);
  const chi2_red = chi2 / dof;

  return {
    rex: Math.max(0, rex_raw),
    rex_raw,
    rex_err,
    chi2_red,
    is_flat: chi2_red < 1.5,
    weighted_mean_r2,
  };
}

export function estimateDeltaOmegaFastExchange(
  rex: number,
  kex: number,
  pb: number,
  b0_mhz: number,
  xi_ratio: number = 0.101329118
): number {
  if (rex <= 0 || kex <= 0 || pb <= 0 || pb >= 1.0 || b0_mhz <= 0) {
    return 0.0;
  }

  const pa = 1.0 - pb;
  const denom = pa * pb;
  if (denom <= 1e-9) return 0.0;

  const dw_rad_s = Math.sqrt((rex * kex) / denom);
  const nu0_mhz = b0_mhz * xi_ratio;
  const dw_ppm = dw_rad_s / (2.0 * Math.PI * nu0_mhz);
  return Number(dw_ppm.toFixed(3));
}
